File: db/scripts/build_schema_full.py
#!/usr/bin/env python3

"""Build bootstrap schema SQL from db/migrations/*.sql.

Outputs:
- db/schema_full.sql
- (optional) db/schema_core.sql, db/schema_auth.sql, db/schema_scam.sql, db/schema_social.sql

Behavior:
- Sorts migrations numerically by prefix: 1.2 < 1.10 < 1.20
- Concatenates with section headers per migration
- Strips per-migration BEGIN/COMMIT/ROLLBACK/START TRANSACTION lines (line-only)
  to avoid nested transactions, then wraps final output in single BEGIN/COMMIT
- Tries to reduce duplicate-object errors by upgrading:
    CREATE EXTENSION -> CREATE EXTENSION IF NOT EXISTS
    CREATE TABLE     -> CREATE TABLE IF NOT EXISTS
    CREATE INDEX     -> CREATE INDEX IF NOT EXISTS
  (only when the statement doesn't already include IF NOT EXISTS)

This script uses only the Python stdlib.
"""

from __future__ import annotations

import argparse
import os
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable, List, Tuple, Dict


@dataclass(frozen=True)
class Migration:
    file_name: str
    abs_path: Path
    rel_path: str
    version_raw: str
    version_parts: Tuple[int, ...]
    category: str  # core|auth|scam|social


TXN_LINE_RE = re.compile(r"^\s*(BEGIN|COMMIT|ROLLBACK|START\s+TRANSACTION)\s*;\s*$", re.IGNORECASE)

CREATE_EXT_RE = re.compile(r"(^\s*CREATE\s+EXTENSION\s+)(?!IF\s+NOT\s+EXISTS\b)", re.IGNORECASE | re.MULTILINE)
CREATE_TABLE_RE = re.compile(r"(^\s*CREATE\s+TABLE\s+)(?!IF\s+NOT\s+EXISTS\b)", re.IGNORECASE | re.MULTILINE)
CREATE_INDEX_RE = re.compile(r"(^\s*CREATE\s+(?:UNIQUE\s+)?INDEX\s+)(?!IF\s+NOT\s+EXISTS\b)", re.IGNORECASE | re.MULTILINE)

CONCURRENTLY_RE = re.compile(r"\bCREATE\s+INDEX\s+CONCURRENTLY\b", re.IGNORECASE)


def parse_version(file_name: str) -> Tuple[str, Tuple[int, ...]]:
    base = os.path.basename(file_name)
    prefix = base.split("__", 1)[0]
    parts: List[int] = []
    for seg in prefix.split("."):
        seg = seg.strip()
        if not seg:
            continue
        try:
            parts.append(int(seg))
        except ValueError:
            # keep parsing best-effort; non-numeric segments just omitted
            pass
    return prefix, tuple(parts)


def version_key(parts: Tuple[int, ...]) -> Tuple[int, ...]:
    # Compare lexicographically, treating missing segments as 0.
    # We normalize to a fixed length to avoid weird comparisons.
    max_len = 4
    padded = list(parts[:max_len]) + [0] * (max_len - len(parts))
    return tuple(padded)


def classify_category(file_name: str) -> str:
    lower = file_name.lower()

    if re.search(r"(scam|fraud|phone|bank|money_mule|mule)", lower):
        return "scam"
    if re.search(r"(user|users|session|sessions|role|roles|auth|password|reset|verify|email)", lower):
        return "auth"
    if re.search(
        r"(post|posts|comment|comments|bookmark|bookmarks|message|messages|chat|dm|notification|notifications|social|file|files|image|images)",
        lower,
    ):
        return "social"
    return "core"


def strip_per_migration_transactions(sql: str) -> Tuple[str, bool]:
    lines = sql.splitlines()
    out: List[str] = []
    stripped = False
    for line in lines:
        if TXN_LINE_RE.match(line):
            stripped = True
            continue
        out.append(line)
    return "\n".join(out) + ("\n" if sql.endswith("\n") else ""), stripped


def upgrade_if_not_exists(sql: str) -> Tuple[str, bool]:
    changed = False

    def _ext(m: re.Match[str]) -> str:
        nonlocal changed
        changed = True
        return m.group(1) + "IF NOT EXISTS "

    def _tbl(m: re.Match[str]) -> str:
        nonlocal changed
        changed = True
        return m.group(1) + "IF NOT EXISTS "

    def _idx(m: re.Match[str]) -> str:
        nonlocal changed
        changed = True
        return m.group(1) + "IF NOT EXISTS "

    sql2 = CREATE_EXT_RE.sub(_ext, sql)
    sql2 = CREATE_TABLE_RE.sub(_tbl, sql2)
    sql2 = CREATE_INDEX_RE.sub(_idx, sql2)

    return sql2, changed


def section_header(file_name: str) -> str:
    return (
        "-- =====================================================\n"
        f"-- MIGRATION: {file_name}\n"
        "-- =====================================================\n\n"
    )


def file_header(out_rel: str, generated_at: str, included: List[str]) -> str:
    lines = [
        "-- =============================================",
        "-- AUTO-GENERATED SCHEMA FILE (DO NOT EDIT)",
        f"-- Output: {out_rel}",
        f"-- GeneratedAt: {generated_at}",
        "-- Included migrations:",
    ]
    lines.extend([f"--  - {f}" for f in included])
    lines.append("-- =============================================")
    return "\n".join(lines) + "\n\n"


def read_utf8(p: Path) -> str:
    return p.read_text(encoding="utf-8")


def write_utf8(p: Path, s: str) -> None:
    p.write_text(s, encoding="utf-8")


def load_migrations(repo_root: Path) -> Tuple[List[Migration], List[str]]:
    mig_dir = repo_root / "db" / "migrations"
    if not mig_dir.exists():
        raise RuntimeError(f"Missing migrations dir: {mig_dir}")

    warnings: List[str] = []
    file_names = [f.name for f in mig_dir.iterdir() if f.is_file() and f.name.lower().endswith(".sql")]
    if not file_names:
        raise RuntimeError(f"No .sql files found in {mig_dir}")

    prefix_map: Dict[str, List[str]] = {}
    migrations: List[Migration] = []

    for fn in file_names:
        abs_path = mig_dir / fn
        rel_path = str(Path("db") / "migrations" / fn).replace(os.sep, "/")
        version_raw, parts = parse_version(fn)
        cat = classify_category(fn)

        prefix_map.setdefault(version_raw, []).append(rel_path)
        migrations.append(
            Migration(
                file_name=fn,
                abs_path=abs_path,
                rel_path=rel_path,
                version_raw=version_raw,
                version_parts=parts,
                category=cat,
            )
        )

    for pref, files in prefix_map.items():
        if len(files) > 1:
            warnings.append(f"Duplicate migration prefix '{pref}': {', '.join(files)}")

    migrations.sort(key=lambda m: (version_key(m.version_parts), m.file_name))
    return migrations, warnings


def build_schema_file(out_path: Path, out_rel: str, migrations: List[Migration]) -> Tuple[List[str], int, int]:
    warnings: List[str] = []
    stripped_count = 0
    upgraded_count = 0

    generated_at = datetime.now(timezone.utc).isoformat()
    included = [m.file_name for m in migrations]

    parts: List[str] = []
    parts.append(file_header(out_rel=out_rel, generated_at=generated_at, included=included))
    parts.append("BEGIN;\n\n")

    for m in migrations:
        raw = read_utf8(m.abs_path)

        if CONCURRENTLY_RE.search(raw):
            warnings.append(
                f"Found CREATE INDEX CONCURRENTLY in {m.rel_path}; this cannot run inside a transaction (BEGIN/COMMIT)."
            )

        stripped_sql, stripped = strip_per_migration_transactions(raw)
        if stripped:
            stripped_count += 1

        upgraded_sql, upgraded = upgrade_if_not_exists(stripped_sql)
        if upgraded:
            upgraded_count += 1

        parts.append(section_header(m.file_name))
        parts.append(upgraded_sql if upgraded_sql.endswith("\n") else upgraded_sql + "\n")
        parts.append("\n")

    parts.append("COMMIT;\n")

    write_utf8(out_path, "".join(parts))
    return warnings, stripped_count, upgraded_count


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--categorized", action="store_true", help="Also emit categorized schema files")
    args = ap.parse_args()

    repo_root = Path(__file__).resolve().parents[1]
    migrations, warnings = load_migrations(repo_root)

    out_full = repo_root / "db" / "schema_full.sql"

    print(f"[schema] repo_root: {repo_root}")
    print(f"[schema] migrations: {len(migrations)}")
    print(f"[schema] out: {out_full}")

    w2, stripped_count, upgraded_count = build_schema_file(
        out_path=out_full, out_rel="db/schema_full.sql", migrations=migrations
    )
    warnings.extend(w2)

    if args.categorized:
        by_cat: Dict[str, List[Migration]] = {"core": [], "auth": [], "scam": [], "social": []}
        for m in migrations:
            by_cat[m.category].append(m)

        cat_outs = [
            ("core", repo_root  / "schema_core.sql"),
            ("auth", repo_root / "schema_auth.sql"),
            ("scam", repo_root / "schema_scam.sql"),
            ("social", repo_root / "schema_social.sql"),
        ]

        for cat, outp in cat_outs:
            if not by_cat[cat]:
                continue
            print(f"[schema] out: {outp} ({cat}, {len(by_cat[cat])} migrations)")
            w_cat, _, _ = build_schema_file(out_path=outp, out_rel=f"schema_{cat}.sql", migrations=by_cat[cat])
            warnings.extend([f"schema_{cat}.sql: {w}" for w in w_cat])

    print(f"[schema] stripped per-migration TXN wrappers: {stripped_count}")
    print(f"[schema] upgraded CREATE .. IF NOT EXISTS (files changed): {upgraded_count}")

    if warnings:
        print("[schema] warnings:")
        for w in warnings:
            print(f"  - {w}")

    print("[schema] done")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: db/scripts/build-schema-full.ts
#!/usr/bin/env node

/**
 * Build bootstrap schema SQL from db/migrations/*.sql
 *
 * Outputs:
 * - db/schema_full.sql
 * - (optional) db/schema_core.sql, db/schema_auth.sql, db/schema_scam.sql, db/schema_social.sql
 *
 * Notes:
 * - Preserves migration order via numeric prefix sort (1.2 < 1.10 < 1.20)
 * - Keeps SQL mostly verbatim
 * - Removes per-migration BEGIN/COMMIT wrappers (line-only) then wraps the final output in one BEGIN/COMMIT
 * - Tries to reduce duplicate-object errors by upgrading CREATE EXTENSION/TABLE/INDEX to IF NOT EXISTS when safe
 */

import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

type Category = "core" | "auth" | "scam" | "social";

type Migration = {
  fileName: string;
  absPath: string;
  relPath: string;
  versionRaw: string;
  versionParts: number[];
  category: Category;
};

function toPosix(p: string): string {
  return p.split(path.sep).join("/");
}

function isSqlFileName(name: string): boolean {
  return name.toLowerCase().endsWith(".sql");
}

function parseMigrationVersion(fileName: string): { versionRaw: string; versionParts: number[] } {
  const base = path.basename(fileName);
  const idx = base.indexOf("__");
  const raw = idx >= 0 ? base.slice(0, idx) : base.replace(/\.sql$/i, "");
  const parts = raw
    .split(".")
    .map((s) => s.trim())
    .filter(Boolean)
    .map((s) => Number.parseInt(s, 10))
    .filter((n) => Number.isFinite(n));
  return { versionRaw: raw, versionParts: parts };
}

function compareVersionParts(a: number[], b: number[]): number {
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const av = Number.isFinite(a[i]) ? a[i] : 0;
    const bv = Number.isFinite(b[i]) ? b[i] : 0;
    if (av !== bv) return av - bv;
  }
  return 0;
}

function classifyCategory(fileName: string): Category {
  const lower = fileName.toLowerCase();

  // scam domain
  if (/(scam|fraud|phone|bank|money_mule|mule)/i.test(lower)) return "scam";

  // auth / identity
  if (/(user|users|session|sessions|role|roles|auth|password|reset|verify|email)/i.test(lower)) return "auth";

  // social / content / messaging
  if (/(post|posts|comment|comments|bookmark|bookmarks|message|messages|chat|dm|notification|notifications|social|file|files|image|images)/i.test(lower)) {
    return "social";
  }

  return "core";
}

async function readUtf8(absPath: string): Promise<string> {
  const buf = await fs.readFile(absPath);
  return buf.toString("utf8");
}

function ensureTrailingNewline(s: string): string {
  return s.endsWith("\n") ? s : `${s}\n`;
}

function stripPerMigrationTransactions(sql: string): { sql: string; stripped: boolean } {
  // Line-based only: remove standalone BEGIN;/COMMIT;/ROLLBACK;/START TRANSACTION;
  // This avoids nested transactions when we wrap the final output.
  const lines = sql.split(/\r?\n/);
  let stripped = false;
  const out: string[] = [];
  for (const line of lines) {
    if (/^\s*(BEGIN|COMMIT|ROLLBACK|START\s+TRANSACTION)\s*;\s*$/i.test(line)) {
      stripped = true;
      continue;
    }
    out.push(line);
  }
  return { sql: out.join("\n"), stripped };
}

function upgradeCreateIfNotExists(sql: string): { sql: string; changed: boolean } {
  let changed = false;
  const before = sql;

  // CREATE EXTENSION IF NOT EXISTS
  sql = sql.replace(/(^\s*CREATE\s+EXTENSION\s+)(?!IF\s+NOT\s+EXISTS\b)/gim, (_m, p1: string) => {
    changed = true;
    return `${p1}IF NOT EXISTS `;
  });

  // CREATE TABLE IF NOT EXISTS
  sql = sql.replace(/(^\s*CREATE\s+TABLE\s+)(?!IF\s+NOT\s+EXISTS\b)/gim, (_m, p1: string) => {
    changed = true;
    return `${p1}IF NOT EXISTS `;
  });

  // CREATE [UNIQUE] INDEX IF NOT EXISTS
  sql = sql.replace(/(^\s*CREATE\s+(?:UNIQUE\s+)?INDEX\s+)(?!IF\s+NOT\s+EXISTS\b)/gim, (_m, p1: string) => {
    changed = true;
    return `${p1}IF NOT EXISTS `;
  });

  if (!changed) return { sql: before, changed: false };
  return { sql, changed };
}

function sectionHeader(fileName: string): string {
  return (
    "-- =====================================================\n" +
    `-- MIGRATION: ${fileName}\n` +
    "-- =====================================================\n\n"
  );
}

function fileHeader(opts: { outRel: string; generatedAt: string; included: string[] }): string {
  const lines: string[] = [];
  lines.push("-- =============================================");
  lines.push("-- AUTO-GENERATED SCHEMA FILE (DO NOT EDIT)");
  lines.push(`-- Output: ${opts.outRel}`);
  lines.push(`-- GeneratedAt: ${opts.generatedAt}`);
  lines.push("-- Included migrations:");
  for (const f of opts.included) lines.push(`--  - ${f}`);
  lines.push("-- =============================================");
  return `${lines.join("\n")}\n\n`;
}

async function writeUtf8(absPath: string, content: string): Promise<void> {
  await fs.writeFile(absPath, content, "utf8");
}

function parseArgs(argv: string[]): { categorized: boolean } {
  return {
    categorized: argv.includes("--categorized"),
  };
}

async function buildOne(opts: {
  repoRoot: string;
  outAbs: string;
  outRel: string;
  generatedAt: string;
  migrations: Migration[];
}): Promise<{ warnings: string[]; strippedCount: number; upgradedCount: number } > {
  const warnings: string[] = [];
  let strippedCount = 0;
  let upgradedCount = 0;

  const included = opts.migrations.map((m) => m.fileName);
  const parts: string[] = [];

  parts.push(fileHeader({ outRel: opts.outRel, generatedAt: opts.generatedAt, included }));
  parts.push("BEGIN;\n\n");

  for (const m of opts.migrations) {
    const raw = await readUtf8(m.absPath);
    const stripped = stripPerMigrationTransactions(raw);
    if (stripped.stripped) strippedCount++;

    const upgraded = upgradeCreateIfNotExists(stripped.sql);
    if (upgraded.changed) upgradedCount++;

    if (/\bCREATE\s+INDEX\s+CONCURRENTLY\b/i.test(raw)) {
      warnings.push(`Found CREATE INDEX CONCURRENTLY in ${m.relPath}; this cannot run inside a transaction.`);
    }

    parts.push(sectionHeader(m.fileName));
    parts.push(ensureTrailingNewline(upgraded.sql));
    parts.push("\n");
  }

  parts.push("COMMIT;\n");

  await writeUtf8(opts.outAbs, parts.join(""));

  return { warnings, strippedCount, upgradedCount };
}

async function main(): Promise<void> {
  const { categorized } = parseArgs(process.argv.slice(2));

  const __filename = fileURLToPath(import.meta.url);
  const __dirname = path.dirname(__filename);
  const repoRoot = path.resolve(__dirname, "..");

  const migrationsDir = path.join(repoRoot, "db", "migrations");
  const outFullAbs = path.join(repoRoot, "db", "schema_full.sql");

  const generatedAt = new Date().toISOString();

  const entries = await fs.readdir(migrationsDir, { withFileTypes: true });
  const fileNames = entries
    .filter((e) => e.isFile() && isSqlFileName(e.name))
    .map((e) => e.name);

  if (fileNames.length === 0) {
    throw new Error(`No .sql files found in ${migrationsDir}`);
  }

  const prefixToFiles = new Map<string, string[]>();

  const migrations: Migration[] = fileNames.map((fileName) => {
    const absPath = path.join(migrationsDir, fileName);
    const relPath = toPosix(path.join("db", "migrations", fileName));
    const { versionRaw, versionParts } = parseMigrationVersion(fileName);
    const category = classifyCategory(fileName);

    const arr = prefixToFiles.get(versionRaw) ?? [];
    arr.push(relPath);
    prefixToFiles.set(versionRaw, arr);

    return { fileName, absPath, relPath, versionRaw, versionParts, category };
  });

  const warnings: string[] = [];
  for (const [prefix, files] of prefixToFiles.entries()) {
    if (files.length > 1) warnings.push(`Duplicate migration prefix '${prefix}': ${files.join(", ")}`);
  }

  migrations.sort((a, b) => {
    const c = compareVersionParts(a.versionParts, b.versionParts);
    if (c !== 0) return c;
    return a.fileName.localeCompare(b.fileName, "en");
  });

  console.log(`[schema] migrations: ${migrations.length}`);
  console.log(`[schema] out: ${outFullAbs}`);

  const fullRes = await buildOne({
    repoRoot,
    outAbs: outFullAbs,
    outRel: "db/schema_full.sql",
    generatedAt,
    migrations,
  });

  warnings.push(...fullRes.warnings);

  if (categorized) {
    const byCat: Record<Category, Migration[]> = {
      core: [],
      auth: [],
      scam: [],
      social: [],
    };

    for (const m of migrations) byCat[m.category].push(m);

    const outMap: Array<{ cat: Category; out: string }> = [
      { cat: "core", out: "schema_core.sql" },
      { cat: "auth", out: "schema_auth.sql" },
      { cat: "scam", out: "schema_scam.sql" },
      { cat: "social", out: "schema_social.sql" },
    ];

    for (const { cat, out } of outMap) {
      if (byCat[cat].length === 0) continue;
      const outAbs = path.join(repoRoot, out);
      console.log(`[schema] out: ${outAbs} (${cat}, ${byCat[cat].length} migrations)`);
      const res = await buildOne({ repoRoot, outAbs, outRel: out, generatedAt, migrations: byCat[cat] });
      warnings.push(...res.warnings.map((w) => `${out}: ${w}`));
    }
  }

  console.log(`[schema] stripped per-migration TXN wrappers: ${fullRes.strippedCount}`);
  console.log(`[schema] upgraded CREATE .. IF NOT EXISTS (files changed): ${fullRes.upgradedCount}`);

  if (warnings.length > 0) {
    console.warn("[schema] warnings:");
    for (const w of warnings) console.warn(`  - ${w}`);
  }

  console.log("[schema] done");
}

main().catch((err: unknown) => {
  const msg = err instanceof Error ? err.stack || err.message : String(err);
  console.error("[schema] fatal:", msg);
  process.exitCode = 1;
});
